### PR TITLE
Fix `undefined-loop-variable` with `NoReturn` and `Never`

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Write a good description on what the PR does.
- [x] Create a news fragment with `towncrier create <IssueNumber>.<type>` which will be
  included in the changelog. `<type>` can be one of: new_check, removed_check, extension,
  false_positive, false_negative, bugfix, other, internal. If necessary you can write
  details or offer examples on how the new change is supposed to work.
- [x] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Continuation of https://github.com/PyCQA/pylint/pull/7419

@DetachHead Sorry not sure what went wrong here. I tried to push my fix to your branch but it closed the PR... Sorry about that!

This should work 😄 

Tested on 3.11 locally as well.

Closes https://github.com/PyCQA/pylint/issues/7311

### CODE REVIEW

File: pylint/checkers/variables.py [hunk 1]
```diff
@@ -2254,13 +2259,25 @@ def _loopvar_name(self, node: astroid.Name) -> None:
...
+        for else_stmt in assign.orelse:
+            if isinstance(
                 else_stmt, (nodes.Return, nodes.Raise, nodes.Break, nodes.Continue)
-            )
-            for else_stmt in assign.orelse
-        ):
-            return
+            ):
+                return
+            # TODO: 2.16: Consider using RefactoringChecker._is_function_def_never_returning
```
Comment (DanielNoord) [inline pylint/checkers/variables.py:2267]: This function doesn't actually infer but uses the `attrname`, which we also match `not_typing.we_do_return.NoReturn`. That isn't really safe..

File: tests/functional/u/undefined/undefined_loop_variable.py [hunk 2]
```diff
@@ -1,9 +1,12 @@
 # pylint: disable=missing-docstring,redefined-builtin, consider-using-f-string, unnecessary-direct-lambda-call
 
+from typing import NoReturn
+
+
 def do_stuff(some_random_list):
     for var in some_random_list:
         pass
-    return var # [undefined-loop-variable]
+    return var  # [undefined-loop-variable]
```
Comment (DanielNoord) [inline tests/functional/u/undefined/undefined_loop_variable.py:9]: I let `black` reformat this file a bit. I find it really annoying when auto-formatting doesn't work on these test files 😅

File: doc/whatsnew/fragments/7311.false_positive [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Fix false positive for ``undefined-loop-variable`` in ``for-else`` loops that use a function that is
+typed as being ``NoReturn`` or ``Never``.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/7311.false_positive:2]:
```suggestion
Fix false positive for ``undefined-loop-variable`` in ``for-else`` loops that use a function
having a return type annotation of ``NoReturn`` or ``Never``.
```

File: tests/functional/u/undefined/undefined_loop_variable.py [hunk 4]
```diff
@@ -125,25 +130,33 @@ def for_else_continue(iterable):
...
 for item in lst:
     pass
 
-bigger = [
-    [
-        x for x in lst2 if x > item
-    ]
-    for item in lst
-]
+bigger = [[x for x in lst2 if x > item] for item in lst]
```
Comment (jacobtylerwalls) [inline tests/functional/u/undefined/undefined_loop_variable.py:151]: Do you mind reverting this change? The variable checker depends on line numbers, so changing the line spacing can change the result.

File: tests/functional/u/undefined/undefined_loop_variable.py (hunk 2 above)
Comment (jacobtylerwalls) [reply tests/functional/u/undefined/undefined_loop_variable.py:9]: I understand, but see below for a concern about the variables checker. In general I'd prefer not to do formatting fixes on test files.

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]

File: tests/functional/u/undefined/undefined_loop_variable.py (hunk 4 above)
Comment (DanielNoord) [reply tests/functional/u/undefined/undefined_loop_variable.py:151]: Good catch! I'll revert!

**Review by jacobtylerwalls — APPROVED** [review]
Looks good!

**Review by Pierre-Sassoulas — DISMISSED** [review]
👍

**Review by jacobtylerwalls — APPROVED** [review]

**Comment by jacobtylerwalls** [comment]
Do you know what's up with the test failure?

**Comment by DanielNoord** [comment]
Probably incorrect checking of `typing_extensions`. I'll have a look on Monday.

**Comment by DanielNoord** [comment]
`coveralls` seems incorrect here..

**Review by Pierre-Sassoulas — APPROVED** [review]
Let's ship it 🎉

**Pull request merged**